crates/ruff_db/src/file_system.rs

```diff
@@ -85,6 +85,245 @@ impl FileSystemPath {
+    /// assert!(path.starts_with("/etc/passwd"));
+    /// assert!(path.starts_with("/etc/passwd/")); // extra slash is okay
+    /// assert!(path.starts_with("/etc/passwd///")); // multiple extra slashes are okay
+    ///
+    /// assert!(!path.starts_with("/e"));
+    /// assert!(!path.starts_with("/etc/passwd.txt"));
+    ///
+    /// assert!(!FileSystemPath::new("/etc/foo.rs").starts_with("/etc/foo"));
+    /// ```
+    #[inline]
+    #[must_use]
+    pub fn starts_with(&self, base: impl AsRef<FileSystemPath>) -> bool {
+        self.0.starts_with(base.as_ref())
+    }
+
+    /// Determines whether `child` is a suffix of `self`.
+    ///
+    /// Only considers whole path components to match.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ruff_db::file_system::FileSystemPath;
+    ///
+    /// let path = FileSystemPath::new("/etc/resolv.conf");
+    ///
+    /// assert!(path.ends_with("resolv.conf"));
+    /// assert!(path.ends_with("etc/resolv.conf"));
+    /// assert!(path.ends_with("/etc/resolv.conf"));
+    ///
+    /// assert!(!path.ends_with("/resolv.conf"));
+    /// assert!(!path.ends_with("conf")); // use .extension() instead
+    /// ```
+    #[inline]
+    #[must_use]
+    pub fn ends_with(&self, child: impl AsRef<FileSystemPath>) -> bool {
+        self.0.ends_with(child.as_ref())
+    }
+
+    /// Returns the `FileSystemPath` without its final component, if there is one.
+    ///
+    /// Returns [`None`] if the path terminates in a root or prefix.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ruff_db::file_system::FileSystemPath;
+    ///
+    /// let path = FileSystemPath::new("/foo/bar");
+    /// let parent = path.parent().unwrap();
+    /// assert_eq!(parent, FileSystemPath::new("/foo"));
+    ///
+    /// let grand_parent = parent.parent().unwrap();
+    /// assert_eq!(grand_parent, FileSystemPath::new("/"));
+    /// assert_eq!(grand_parent.parent(), None);
+    /// ```
+    #[inline]
+    #[must_use]
+    pub fn parent(&self) -> Option<&FileSystemPath> {
+        self.0.parent().map(FileSystemPath::new)
+    }
+
+    /// Produces an iterator over the [`camino::Utf8Component`]s of the path.
+    ///
+    /// When parsing the path, there is a small amount of normalization:
+    ///
+    /// * Repeated separators are ignored, so `a/b` and `a//b` both have
+    ///   `a` and `b` as components.
+    ///
+    /// * Occurrences of `.` are normalized away, except if they are at the
+    ///   beginning of the path. For example, `a/./b`, `a/b/`, `a/b/.` and
+    ///   `a/b` all have `a` and `b` as components, but `./a/b` starts with
+    ///   an additional [`CurDir`] component.
+    ///
+    /// * A trailing slash is normalized away, `/a/b` and `/a/b/` are equivalent.
+    ///
+    /// Note that no other normalization takes place; in particular, `a/c`
+    /// and `a/b/../c` are distinct, to account for the possibility that `b`
+    /// is a symbolic link (so its parent isn't `a`).
```

Comment:
I need to add a to-do list item to check in what situations Python normalizes the import statement itself. E.g. `import foo.bar.baz` is obviously normalized to the same thing as `..bar.baz` and `.baz` in some situations, and I don't know how symlinks play into that. But I think it's correct that we shouldn't normalize the path here (the normalization of the import statement should probably happen at an earlier stage)